core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,391 @@
+   *
+   * @param localBlockStore the local block store
+   */
+  public UnderFileSystemBlockStore(BlockStore localBlockStore) {
+    mLocalBlockStore = localBlockStore;
+  }
+
+  /**
+   * Acquires access for a UFS block given a {@link UnderFileSystemBlockMeta} and the limit on
+   * the maximum concurrency on the block.
+   *
+   * @param sessionId the session ID
+   * @param blockId maximum concurrency
+   * @param options the options
+   * @return whether an access token is acquired
+   * @throws BlockAlreadyExistsException if the block already exists for a session ID
+   */
+  public boolean acquireAccess(long sessionId, long blockId, OpenUfsBlockOptions options)
+      throws BlockAlreadyExistsException {
+    UnderFileSystemBlockMeta blockMeta = new UnderFileSystemBlockMeta(sessionId, blockId, options);
+    mLock.lock();
+    try {
+      Key key = new Key(sessionId, blockId);
+      if (mBlocks.containsKey(key)) {
+        throw new BlockAlreadyExistsException(ExceptionMessage.UFS_BLOCK_ALREADY_EXISTS_FOR_SESSION,
+            blockId, blockMeta.getUnderFileSystemPath(), sessionId);
+      }
+      Set<Long> sessionIds = mBlockIdToSessionIds.get(blockId);
+      if (sessionIds != null && sessionIds.size() >= options.getMaxUfsReadConcurrency()) {
+        return false;
+      }
+      if (sessionIds == null) {
+        sessionIds = new HashSet<>();
+        mBlockIdToSessionIds.put(blockId, sessionIds);
+      }
+      sessionIds.add(sessionId);
+
+      mBlocks.put(key, new BlockInfo(blockMeta));
+
+      Set<Long> blockIds = mSessionIdToBlockIds.get(sessionId);
+      if (blockIds == null) {
+        blockIds = new HashSet<>();
+        mSessionIdToBlockIds.put(sessionId, blockIds);
+      }
+      blockIds.add(blockId);
+    } finally {
+      mLock.unlock();
+    }
+    return true;
+  }
+
+  /**
+   * Cleans up the block reader or writer and checks whether it is necessary to commit the block
+   * to Local block store.
+   *
+   * During UFS block read, this is triggered when the block is unlocked.
+   * During UFS block write, this is triggered when the UFS block is committed.
+   *
+   * @param sessionId the session ID
+   * @param blockId the block ID
+   * @return true if block is to be committed into Local block store
+   * @throws IOException if it fails to clean up
+   */
+  public boolean cleanup(long sessionId, long blockId) throws IOException {
+    BlockInfo blockInfo;
+    mLock.lock();
+    try {
+      blockInfo = mBlocks.get(new Key(sessionId, blockId));
+      if (blockInfo == null) {
+        return false;
+      }
+    } finally {
+      mLock.unlock();
+    }
+    return blockInfo.closeReaderOrWriter();
+  }
+
+  /**
+   * Releases the access token of this block by removing this (sessionId, blockId) pair from the
+   * store.
+   *
+   * @param sessionId the session ID
+   * @param blockId the block ID
+   */
+  public void releaseAccess(long sessionId, long blockId) {
```

Comment:
@peisun1115 is it necessary to throw exceptions if `sessionId` and `blockId` combination is not found?